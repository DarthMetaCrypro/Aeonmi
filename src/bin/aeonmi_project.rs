// Tiny compatibility wrapper binary.
// Reuses your real main so tests & old scripts keep working.
include!("../main.rs");
