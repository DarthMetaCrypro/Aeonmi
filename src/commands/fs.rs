use anyhow::Result;
use std::path::PathBuf;
use std::fs;
use std::io::Write;
use chrono::Local;

pub fn new_file(path: Option<PathBuf>) -> Result<()> {
    let target = path.unwrap_or_else(|| PathBuf::from("untitled.ai"));
    if target.exists() {
        println!("new: '{}' already exists (leaving unchanged)", target.display());
        return Ok(());
    }
    if let Some(parent) = target.parent() { if !parent.as_os_str().is_empty() { fs::create_dir_all(parent)?; } }
    let now = Local::now().format("%Y-%m-%d %H:%M:%S");
    let template = format!(r#"// Aeonmi source created {now}
let greeting = "Hello Aeonmi";
function square(x) {{ return x * x; }}
let total = 0;
for let i = 0; i < 5; i = i + 1 {{ total = total + square(i); }}
log(greeting, total);
"#);
    let mut f = fs::File::create(&target)?;
    f.write_all(template.as_bytes())?;
    println!("new: created '{}'", target.display());
    Ok(())
}

pub fn open(path: PathBuf) -> Result<()> {
    println!("open: '{}' (placeholder)", path.display());
    Ok(())
}

pub fn save(path: Option<PathBuf>) -> Result<()> {
    if let Some(p) = path {
        println!("save: '{}' (placeholder)", p.display());
    } else {
        println!("save: current buffer (placeholder)");
    }
    Ok(())
}

pub fn save_as(path: PathBuf) -> Result<()> {
    println!("saveas: '{}' (placeholder)", path.display());
    Ok(())
}

pub fn close(path: Option<PathBuf>) -> Result<()> {
    if let Some(p) = path {
        println!("close: '{}' (placeholder)", p.display());
    } else {
        println!("close: current buffer (placeholder)");
    }
    Ok(())
}

pub fn import(path: PathBuf) -> Result<()> {
    println!("import: '{}' (placeholder)", path.display());
    Ok(())
}

pub fn export(path: PathBuf, format: Option<String>) -> Result<()> {
    println!("export: '{}' as '{}' (placeholder)", path.display(), format.unwrap_or_else(|| "auto".into()));
    Ok(())
}

pub fn upload(path: PathBuf) -> Result<()> {
    use std::fs;
    use std::env;

    let src = if path.is_absolute() { path } else { env::current_dir()?.join(path) };
    if !src.exists() {
        anyhow::bail!("upload failed: source '{}' does not exist", src.display());
    }

    let repo_root = env::current_dir()?;
    let uploads_dir = repo_root.join("uploads");
    if !uploads_dir.exists() {
        fs::create_dir_all(&uploads_dir)?;
    }

    if src.is_dir() {
        let options = fs_extra::dir::CopyOptions::new();
        let to = uploads_dir.join(src.file_name().unwrap_or_else(|| std::ffi::OsStr::new("uploaded_dir")));
        fs_extra::dir::copy(&src, &to, &options)?;
        println!("uploaded dir '{}' -> '{}'", src.display(), to.display());
        return Ok(());
    }

    let filename = src
        .file_name()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_else(|| "uploaded.bin".into());
    let dst = uploads_dir.join(filename);
    fs::copy(&src, &dst)?;
    println!("uploaded '{}' -> '{}'", src.display(), dst.display());
    Ok(())
}

pub fn download(path: PathBuf) -> Result<()> {
    use std::fs;
    use std::env;

    let repo_root = env::current_dir()?;
    let src = if path.is_absolute() { path } else { repo_root.join(&path) };
    if !src.exists() {
        anyhow::bail!("download failed: source '{}' does not exist in workspace", src.display());
    }

    let downloads_dir = repo_root.join("downloads");
    if !downloads_dir.exists() {
        fs::create_dir_all(&downloads_dir)?;
    }

    if src.is_dir() {
        let options = fs_extra::dir::CopyOptions::new();
        let to = downloads_dir.join(src.file_name().unwrap_or_else(|| std::ffi::OsStr::new("downloaded_dir")));
        fs_extra::dir::copy(&src, &to, &options)?;
        println!("downloaded dir '{}' -> '{}'", src.display(), to.display());
        return Ok(());
    }

    let filename = src
        .file_name()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_else(|| "downloaded.bin".into());
    let dst = downloads_dir.join(filename);
    fs::copy(&src, &dst)?;
    println!("downloaded '{}' -> '{}'", src.display(), dst.display());
    Ok(())
}
