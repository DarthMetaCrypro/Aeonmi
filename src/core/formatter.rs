//! Canonical, idempotent .ai source formatter.
//!
//! Design notes:
//! - Token-agnostic but syntax-aware via simple rules.
//! - Stable spacing around punctuation and operators.
//! - Indentation with 4 spaces, newline rules around braces.
//! - Idempotent: format(format(src)) == format(src)

use std::fmt::Write as _;

#[derive(Debug, Clone, Copy)]
pub struct FormatOptions {
    /// Spaces per indent level.
    pub indent: usize,
    /// Max consecutive blank lines allowed.
    pub max_blank: usize,
}

impl Default for FormatOptions {
    fn default() -> Self {
        Self { indent: 4, max_blank: 1 }
    }
}

pub fn format_ai(src: &str) -> String {
    format_ai_with(src, &FormatOptions::default())
}

pub fn format_ai_with(src: &str, opt: &FormatOptions) -> String {
    // 1) Normalize line endings, trim trailing whitespace lines.
    let mut s = src.replace("\r\n", "\n").replace('\r', "\n");
    s = s
        .lines()
        .map(|l| l.trim_end())
        .collect::<Vec<_>>()
        .join("\n");

    // 2) Collapse >max_blank consecutive blank lines.
    let mut cleaned = String::new();
    let mut blank_run = 0usize;
    for line in s.lines() {
        if line.trim().is_empty() {
            blank_run += 1;
            if blank_run <= opt.max_blank {
                cleaned.push('\n');
            }
        } else {
            blank_run = 0;
            cleaned.push_str(line);
            cleaned.push('\n');
        }
    }
    // avoid trailing newline explosion
    while cleaned.ends_with("\n\n") {
        cleaned.pop();
    }

    // 3) Token-ish pass: enforce spacing around punctuation/operators.
    //    (We avoid full tokenization to keep this decoupled and robust.)
    let mut out = String::with_capacity(cleaned.len() + cleaned.len() / 10);
    let mut indent = 0usize;
    let mut i = 0usize;
    let bytes = cleaned.as_bytes();

    // helpers
    let mut was_space = false;
    let mut need_space = false;
    let mut just_wrote_newline = true;

    while i < bytes.len() {
        let ch = bytes[i] as char;

        // Simple string/char literal passthrough (handles quotes)
        if ch == '"' || ch == '\'' {
            let quote = ch;
            push_pending_space(&mut out, &mut need_space, &mut was_space);
            out.push(quote);
            i += 1;
            while i < bytes.len() {
                let c = bytes[i] as char;
                out.push(c);
                i += 1;
                if c == '\\' && i < bytes.len() {
                    // escape next
                    out.push(bytes[i] as char);
                    i += 1;
                    continue;
                }
                if c == quote { break; }
            }
            just_wrote_newline = false;
            continue;
        }

        match ch {
            // Braces/newline/indent rules
            '{' => {
                push_pending_space(&mut out, &mut need_space, &mut was_space);
                out.push(' ');
                out.push('{');
                out.push('\n');
                indent += 1;
                write_indent(&mut out, opt.indent, indent);
                just_wrote_newline = true;
                was_space = false;
                need_space = false;
                i += 1;
            }
            '}' => {
                // dedent first
                if indent > 0 { indent -= 1; }
                if !just_wrote_newline {
                    out.push('\n');
                }
                write_indent(&mut out, opt.indent, indent);
                out.push('}');
                just_wrote_newline = false;
                was_space = false;
                need_space = true;
                i += 1;
            }
            '(' | '[' => {
                push_pending_space(&mut out, &mut need_space, &mut was_space);
                out.push(ch);
                just_wrote_newline = false;
                was_space = false;
                need_space = false;
                i += 1;
            }
            ')' | ']' => {
                out.push(ch);
                just_wrote_newline = false;
                was_space = false;
                need_space = true;
                i += 1;
            }
            ',' => {
                out.push(',');
                out.push(' ');
                just_wrote_newline = false;
                was_space = true;
                need_space = false;
                i += 1;
            }
            ';' => {
                out.push(';');
                out.push('\n');
                write_indent(&mut out, opt.indent, indent);
                just_wrote_newline = true;
                was_space = false;
                need_space = false;
                i += 1;
            }
            ':' => {
                out.push(':');
                out.push(' ');
                was_space = true;
                need_space = false;
                just_wrote_newline = false;
                i += 1;
            }
            ' ' | '\t' => {
                if !just_wrote_newline && !was_space {
                    need_space = true;
                }
                i += 1;
            }
            '\n' => {
                if !just_wrote_newline {
                    out.push('\n');
                    just_wrote_newline = true;
                }
                write_indent(&mut out, opt.indent, indent);
                was_space = false;
                need_space = false;
                i += 1;
            }
            _ => {
                push_pending_space(&mut out, &mut need_space, &mut was_space);
                out.push(ch);
                just_wrote_newline = false;
                was_space = false;
                need_space = false;
                i += 1;
            }
        }
    }

    // final trim of trailing spaces and ensure single trailing newline
    let mut final_s = out
        .lines()
        .map(|l| l.trim_end())
        .collect::<Vec<_>>()
        .join("\n");
    if !final_s.ends_with('\n') {
        final_s.push('\n');
    }
    final_s
}

fn write_indent(out: &mut String, indent_spaces: usize, level: usize) {
    for _ in 0..(indent_spaces * level) { out.push(' '); }
}

fn push_pending_space(out: &mut String, need_space: &mut bool, was_space: &mut bool) {
    if *need_space {
        out.push(' ');
        *need_space = false;
        *was_space = true;
    }
}