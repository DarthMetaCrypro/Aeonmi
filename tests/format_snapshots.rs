//! tests/format_snapshots.rs
//! Snapshot tests for canonical .ai formatting.

use aeonmi_project::core::formatter::format_ai;
use aeonmi_project as _;
use std::fs;
use std::path::PathBuf;

#[test]
fn examples_are_stably_formatted() {
    let examples = [
        "examples/hello.ai",
        "examples/control_flow.ai",
        "examples/functions.ai",
        "examples/glyph.ai",
    ];
    for ex in examples {
        let p = PathBuf::from(ex);
        if let Ok(src) = fs::read_to_string(&p) {
            let out = format_ai(&src);
            insta::assert_snapshot!(p.file_name().unwrap().to_string_lossy().as_ref(), &out);
        }
    }
}
