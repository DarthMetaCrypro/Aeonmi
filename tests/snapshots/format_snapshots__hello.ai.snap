---
source: tests/format_snapshots.rs
expression: "&out"
---
// hello.ai
let x = 42;
log(x);
